selfdrive/ui/replay/filereader.h

```diff
@@ -1,70 +1,91 @@
 #pragma once
 
-#include <thread>
+#include <unordered_map>
+#include <vector>
 
 #include <QElapsedTimer>
 #include <QMultiMap>
 #include <QNetworkAccessManager>
-#include <QReadWriteLock>
 #include <QString>
-#include <QVector>
-#include <QWidget>
+#include <QThread>
 
-#include <bzlib.h>
 #include <capnp/serialize.h>
-#include <kj/io.h>
-#include "cereal/gen/cpp/log.capnp.h"
 
-#include "tools/clib/channel.h"
+#include "cereal/gen/cpp/log.capnp.h"
 
 class FileReader : public QObject {
   Q_OBJECT
 
 public:
-  FileReader(const QString& file_);
-  void startRequest(const QUrl &url);
-  ~FileReader();
-  virtual void readyRead();
-  void httpFinished();
-  virtual void done() {};
+  FileReader(const QString &fn, QObject *parent = nullptr);
+  void read();
+  void abort();
 
-public slots:
-  void process();
-
-protected:
-  QNetworkReply *reply;
+signals:
+  void finished(const QByteArray &dat);
+  void failed(const QString &err);
 
 private:
-  QNetworkAccessManager *qnam;
-  QElapsedTimer timer;
-  QString file;
+  void startHttpRequest();
+  QNetworkReply *reply_ = nullptr;
+  QUrl url_;
+};
+
+enum CameraType {
+  RoadCam = 0,
+  DriverCam,
+  WideRoadCam
+};
+const CameraType ALL_CAMERAS[] = {RoadCam, DriverCam, WideRoadCam};
+const int MAX_CAMERAS = std::size(ALL_CAMERAS);
+
+struct EncodeIdx {
+  int segmentNum;
+  uint32_t frameEncodeId;
 };
 
-typedef QMultiMap<uint64_t, cereal::Event::Reader> Events;
+class Event {
+public:
+  Event(const kj::ArrayPtr<const capnp::word> &amsg) : reader(amsg) {
+    words = kj::ArrayPtr<const capnp::word>(amsg.begin(), reader.getEnd());
+    event = reader.getRoot<cereal::Event>();
+    which = event.which();
+    mono_time = event.getLogMonoTime();
+  }
+  inline kj::ArrayPtr<const capnp::byte> bytes() const { return words.asBytes(); }
+
+  uint64_t mono_time;
+  cereal::Event::Which which;
+  cereal::Event::Reader event;
+  capnp::FlatArrayMessageReader reader;
+  kj::ArrayPtr<const capnp::word> words;
+};
+
+typedef QMultiMap<uint64_t, Event*> Events;
+typedef std::unordered_map<uint32_t, EncodeIdx> EncodeIdxMap;
```

Comment:
not a fan of these typedefs, I think it's pretty confusing when reading "Events" and "EncodeIdxMap" later on